### PR TITLE
Preserve future API mode defaults after customization

### PR DESCRIPTION
**Pull request by PeterDaveHello**
## Summary

- Track the API mode default baseline for customized profiles so future defaults can be appended without restoring defaults that users removed.
- Preserve compatibility for legacy configuration migration and imported backups, including deterministic cleanup of stale API mode state.
- Persist API mode lists, selection state, and baseline metadata atomically, with rollback-safe handling when persistence fails.
- Preserve built-in Azure OpenAI and Ollama rows while materializing and reconciling defaults.

## Validation

- `npm run lint`
- `npm test` (864 tests)
- `npm run build`

## Rollout

This is phase 1 of the future-default rollout. The API mode reorder UI is intentionally deferred until this PR is accepted.

## Summary by CodeRabbit

- **New Features**
  - Added materialized API mode defaults with deterministic default IDs and schema-v2 initialization.
  - Introduced selection-aware API mode config helpers to keep active/custom lists and selection fields in sync.
- **Bug Fixes**
  - Improved schema-v2 migration/import cleanup to sanitize malformed API-mode data and prevent stale defaults or legacy baseline fields from reappearing.
  - Reconciliation now avoids duplicates while preserving provider/custom variants correctly.
- **Tests**
  - Expanded unit coverage for migration, API mode config expansion/update + rollback behavior, import normalization, and reconciliation edge cases.

### CODE REVIEW

File: src/utils/model-name-convert.mjs [hunk 1]
```diff
@@ -219,6 +222,64 @@ export function getApiModesFromConfig(config, onlyActive) {
...
+
+  for (const defaultId of Array.isArray(defaultIds) ? defaultIds : []) {
+    if (knownIdSet.has(defaultId)) continue
+
+    if (defaultId === 'customModel') {
+      knownIds.push(defaultId)
+      knownIdSet.add(defaultId)
+      changed = true
+      continue
+    }
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/utils/model-name-convert.mjs:266]: Not adopting this suggestion. `getApiModesFromConfig` intentionally omits the active `customModel` placeholder, so the explicit baseline entry preserves future-default bookkeeping without materializing that placeholder.

File: src/config/index.mjs [hunk 2]
```diff
@@ -1555,6 +1559,101 @@ function migrateUserConfig(options) {
...
+      activeApiModes: [],
+    },
+    false,
+  )
+  const hasStoredActiveApiModes = Array.isArray(migrated.activeApiModes)
+  const rawKnownApiModeDefaultIds = migrated.knownApiModeDefaultIds
+  const hasCurrentBaseline =
+    Number(migrated.configSchemaVersion) >= 2 &&
+    Array.isArray(rawKnownApiModeDefaultIds) &&
+    (currentDefaultIds.length === 0 || rawKnownApiModeDefaultIds.length > 0) &&
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1688]: Intentionally retained. If a future default catalog is empty, this condition gives it explicit no-op baseline semantics instead of falling into migration.

File: src/utils/model-name-convert.mjs [hunk 3]
```diff
@@ -136,7 +136,10 @@ export function getApiModesFromConfig(config, onlyActive) {
     .filter((apiMode) => {
       if (!apiMode || !apiMode.groupName) return false
       if (AlwaysCustomGroups.includes(apiMode.groupName)) {
-        return Boolean(apiMode.customName && apiMode.customName.trim())
+        return Boolean(
+          (apiMode.customName && apiMode.customName.trim()) ||
+            (!apiMode.isCustom && apiMode.itemName),
+        )
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/utils/model-name-convert.mjs:150]: Fixed in f53bcf1. `AlwaysCustomGroups` rows are now accepted only with a non-empty `customName` or as an exact built-in Azure/Ollama preset, so malformed broad matches are excluded.

File: src/config/index.mjs [hunk 4]
```diff
@@ -1555,6 +1559,101 @@ function migrateUserConfig(options) {
...
+      dirty = true
+    }
+
+    let knownApiModeDefaultIds = hasCurrentBaseline
+      ? canonicalizeModelKeyArray(
+          migrated.knownApiModeDefaultIds
+            .filter((modelName) => typeof modelName === 'string')
+            .map((modelName) => modelName.trim())
+            .filter(Boolean),
+        )
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1737]: Not adopted in this PR. Migration and import are distinct boundaries with intentionally localized behavior, and the tiny duplication does not justify widening the scope.

File: src/config/index.mjs [hunk 5]
```diff
@@ -1555,6 +1559,101 @@ function migrateUserConfig(options) {
...
+        {
+          ...migrationConfig,
+          activeApiModes: activeApiModesForMaterialization,
+          customApiModes,
+        },
+        false,
+      )
+      migrated.activeApiModes = []
+      dirty = true
+    }
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1729]: Fixed in f53bcf1. Migration now filters non-string and blank legacy `activeApiModes` IDs before canonicalization, with migration tests covering malformed entries.

File: src/config/index.mjs [hunk 6]
```diff
@@ -1667,8 +1783,15 @@ export async function getUserConfig() {
...
+    let payloadPersisted = true
     if (Object.keys(payload).length > 0) {
-      await Browser.storage.local.set(payload).catch(() => {})
+      payloadPersisted = await Browser.storage.local
+        .set(payload)
+        .then(() => true)
+        .catch(() => false)
+    }
+    if (payloadPersisted && storageKeysToRemove.length > 0) {
+      await Browser.storage.local.remove(storageKeysToRemove).catch(() => {})
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs]: storageKeysToRemove is populated only in isLiveDefaultProfile. Its queued active/default-baseline keys are non-authoritative invalid sentinels, so a failed remove cannot turn them into a valid active array or schema-v2 baseline; the next getUserConfig read deterministically queues removal again. Commit 15257cc makes this retry intent explicit and adds a fail-once/two-read convergence test verifying both reads return live defaults, the sentinels remain after the first failure, and they are removed on the second read.

File: src/config/index.mjs [hunk 7]
```diff
@@ -1555,6 +1566,101 @@ function migrateUserConfig(options) {
...
+        {
+          ...migrationConfig,
+          activeApiModes: activeApiModesForMaterialization,
+          customApiModes,
+        },
+        false,
+      )
+      migrated.activeApiModes = []
+      dirty = true
+    }
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1729]: Not adopting this. Before migration, defaults were applied after reading storage and were not auto-persisted; the legacy API Modes UI persisted `activeApiModes` only after a list mutation. A stored array is therefore the only surviving customization or snapshot provenance, and equality with today's defaults is ambiguous. Reclassifying it as live could erase explicit customization. Migration coverage intentionally materializes `activeApiModes` against `defaultApiModeIds`.

File: src/config/index.mjs [hunk 8]
```diff
@@ -1555,6 +1566,101 @@ function migrateUserConfig(options) {
...
+
+    if (!hasCurrentBaseline || activeApiModesForMaterialization.length > 0) {
+      customApiModes = getApiModesFromConfig(
+        {
+          ...migrationConfig,
+          activeApiModes: activeApiModesForMaterialization,
+          customApiModes,
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1723]: Acknowledged: materialization snapshots the name, intentionally for a legacy customized profile. Untouched profiles remain live-bound. The current UI does not write `azureDeploymentName` or `ollamaModelName`; it edits names through the API Mode row. Rebinding materialized rows would require provenance metadata to avoid overwriting explicit row edits and is a separate product change.

File: src/popup/sections/ApiModes.jsx [hunk 9]
```diff
@@ -142,16 +143,18 @@ export function ApiModes({ config, updateConfig }) {
...
           apiModeStringArray.includes(config.modelName) &&
           config.modelName !== apiModeToModelName(apiMode)
             ? config.modelName
             : 'customModel',
         apiMode: null,
-      })
+      }
+    }
+    return {}
   }
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/sections/ApiModes.jsx]: Not adopting this. `apiModeStringArray` predates this PR, and both states are updated from the same `nextApiModes` in one `useLayoutEffect`, with no independent mutation or stale path. Removing it is out-of-scope cleanup.

File: src/popup/api-mode-config-utils.mjs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+  return {
+    activeApiModes: currentConfig.activeApiModes,
+    customApiModes: currentConfig.customApiModes,
+    knownApiModeDefaultIds: currentConfig.knownApiModeDefaultIds,
+    modelName: currentConfig.modelName,
+    apiMode: currentConfig.apiMode,
+    ...nextValue,
+  }
+}
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:59]: Not adopting this. The only production caller passes `latestConfigRef.current`, initialized to `defaultConfig` and maintained as an object. Optional chaining would mask an invariant violation and create an incomplete atomic bundle with `undefined` values.

File: src/popup/sections/ApiModes.jsx [hunk 11]
```diff
@@ -232,13 +235,15 @@ export function ApiModes({ config, updateConfig }) {
...
-          : apiModes.map((apiMode, index) => (index === editingIndex ? nextApiMode : apiMode)),
-    }
+    const nextApiModes =
+      editingIndex === -1
+        ? [...apiModes, nextApiMode]
+        : apiModes.map((apiMode, index) => (index === editingIndex ? nextApiMode : apiMode))
+    const selectionPatch =
+      editingIndex !== -1 && isApiModeSelected(apiModes[editingIndex], config)
+        ? { apiMode: nextApiMode }
+        : {}
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/sections/ApiModes.jsx:227]: No change. Updating the selected row via `apiMode` only is pre-existing behavior: `apiMode` is the authoritative structured selection when present, while `modelName` is the legacy fallback. This PR introduced neither that precedence nor a new stale path, and list/selection atomicity remains explicit.

File: src/popup/sections/import-data-cleanup.mjs [hunk 12]
```diff
@@ -10,10 +10,22 @@ const conflictingKeyPairs = [
...
   const keysToRemove = []
 
+  if (apiModeListKeys.some((key) => Object.hasOwn(data, key))) {
+    for (const key of apiModeListKeys) {
+      if (!Object.hasOwn(data, key)) normalizedData[key] = null
+    }
+    for (const key of apiModeSelectionKeys) {
+      if (!Object.hasOwn(data, key)) keysToRemove.push(key)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/sections/import-data-cleanup.mjs:27]: No change. The production import handler parses JSON and rejects `null`, arrays, and non-object values before calling `importDataIntoStorage`/`prepareImportData`, so `data` is guaranteed to be a plain object at this internal boundary.

File: src/config/index.mjs [hunk 13]
```diff
@@ -1037,7 +1041,14 @@ function migrateUserConfig(options) {
...
+    Array.isArray(migrated.activeApiModes) &&
+    migrated.activeApiModes.some((modelName) => typeof modelName !== 'string' || !modelName.trim())
+      ? migrated.activeApiModes.filter(
+          (modelName) => typeof modelName === 'string' && Boolean(modelName.trim()),
+        )
+      : migrated.activeApiModes
+  const canonicalActiveApiModes = canonicalizeModelKeyArray(activeApiModesForCanonicalization)
   if (canonicalActiveApiModes !== migrated.activeApiModes) {
     migrated.activeApiModes = canonicalActiveApiModes
     dirty = true
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1153]: No change. `canonicalizeModelKeyArray` intentionally returns non-array inputs unchanged, preserving the value and reference so `migrateUserConfig` can classify a non-array `activeApiModes` value as an invalid live-default sentinel and remove it rather than materialize it.

File: src/popup/api-mode-config-utils.mjs [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    knownApiModeDefaultIds: currentConfig.knownApiModeDefaultIds,
+    ...(includesSelectionUpdate
+      ? {
+          modelName: currentConfig.modelName,
+          apiMode: currentConfig.apiMode,
+        }
+      : {}),
+    ...nextValue,
+  }
+}
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:59]: No change; this duplicates the earlier resolved suggestion. The sole production caller passes `latestConfigRef.current`, initialized with `defaultConfig` and maintained as an object. Optional chaining would mask an invariant violation and create an incomplete atomic bundle containing `undefined` values.

File: src/popup/api-mode-config-utils.mjs [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+    customApiModes: currentConfig.customApiModes,
+    knownApiModeDefaultIds: currentConfig.knownApiModeDefaultIds,
+    ...(includesSelectionUpdate
+      ? {
+          modelName: currentConfig.modelName,
+          apiMode: currentConfig.apiMode,
+        }
+      : {}),
+    ...nextValue,
+  }
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:58]: No code change is planned because the sole production caller passes `latestConfigRef.current`, which is initialized from `defaultConfig` and maintained as a complete config object. Accepting null or partial input here would manufacture `undefined`/default list payloads and hide an invariant violation. This also duplicates the previously resolved optional-config guard suggestion.

File: src/popup/sections/import-data-cleanup.mjs [hunk 16]
```diff
@@ -10,10 +10,22 @@ const conflictingKeyPairs = [
   ['customClaudeApiUrl', 'customAnthropicApiUrl'],
 ]
 
+const apiModeListKeys = ['activeApiModes', 'customApiModes', 'knownApiModeDefaultIds']
+const apiModeSelectionKeys = ['modelName', 'apiMode']
+
 export function prepareImportData(data) {
   const normalizedData = { ...data }
   const keysToRemove = []
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/sections/import-data-cleanup.mjs:18]: Thanks. No code change is planned here: the sole production import path parses the file, explicitly requires a non-null, non-array object in GeneralPart.jsx, and only then calls importDataIntoStorage/prepareImportData. Treating null or arrays as an empty backup inside this lower-level normalization helper would hide an invalid-backup invariant violation instead of reporting it. This is the same caller-invariant suggestion already assessed in the prior review round.

File: src/popup/api-mode-config-utils.mjs (hunk 15 above)
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:58]: Thanks. No code change is planned here: the sole production caller passes latestConfigRef.current, which is initialized with defaultConfig and maintained as a complete config object across optimistic updates, persistence, rollback, and config loading. Optional chaining here would manufacture undefined fields in an atomic list update and hide an invariant violation rather than recover safely. This duplicates the optional-currentConfig suggestion already assessed in the prior review round.

File: src/popup/sections/import-data-cleanup.mjs (hunk 16 above)
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/sections/import-data-cleanup.mjs:18]: No code change is planned. The sole production import path parses JSON and rejects null, arrays, and non-object values in GeneralPart.jsx before calling importDataIntoStorage/prepareImportData. Treating invalid input as an empty backup here would hide that boundary invariant. This is the same caller-invariant suggestion already assessed in the prior review round.

File: src/popup/api-mode-config-utils.mjs (hunk 15 above)
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:58]: No code change is planned. The sole production caller passes latestConfigRef.current, initialized with defaultConfig and maintained as a complete config object through load, optimistic updates, persistence, and rollback. Falling back to empty arrays or undefined selection fields would hide an invariant violation and could create an incomplete atomic bundle. This duplicates the previously resolved currentConfig guard suggestion.

File: src/utils/model-name-convert.mjs [hunk 17]
```diff
@@ -219,6 +230,69 @@ export function getApiModesFromConfig(config, onlyActive) {
...
+    }
+
+    const materializedModes = getApiModesFromConfig(
+      {
+        ...config,
+        activeApiModes: [defaultId],
+        customApiModes: [],
+      },
+      false,
+    )
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/utils/model-name-convert.mjs:275]: No code change is planned. The only production call is inside migrateUserConfig and passes an object built from migrationConfig plus explicit activeApiModes/customApiModes; all test calls also pass objects. Normalizing null to an empty config here would hide an internal contract violation and reconcile from incomplete provider/config state rather than recover safely.

File: src/config/index.mjs [hunk 18]
```diff
@@ -1132,7 +1136,14 @@ function migrateUserConfig(options) {
...
   }
 
-  const canonicalActiveApiModes = canonicalizeModelKeyArray(migrated.activeApiModes)
+  const activeApiModesForCanonicalization =
+    Array.isArray(migrated.activeApiModes) &&
+    migrated.activeApiModes.some((modelName) => typeof modelName !== 'string' || !modelName.trim())
+      ? migrated.activeApiModes.filter(
+          (modelName) => typeof modelName === 'string' && Boolean(modelName.trim()),
+        )
+      : migrated.activeApiModes
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/config/index.mjs:1149]: Fixed in 8d78eeb. Legacy activeApiModes sanitization now trims string IDs before canonicalization/materialization while preserving the unchanged-array reference fast path. The existing malformed-ID migration test now uses a padded valid ID and verifies both returned config and persisted storage materialize canonical chatgptFree35.

File: src/utils/model-name-convert.mjs [hunk 19]
```diff
@@ -219,6 +230,69 @@ export function getApiModesFromConfig(config, onlyActive) {
...
+    const materializedMode = materializedModes[0]
+    if (!materializedMode) continue
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/utils/model-name-convert.mjs:277]: No code change is planned. An unresolvable default ID indicates that the catalog and model metadata are temporarily inconsistent. Marking it known would permanently suppress that same ID if a later version restores its metadata, whereas leaving it unknown retries safely on the next config read or upgrade and allows self-recovery. The loop still terminates after the finite default list on every call, and all current defaultApiModeIds resolve in tests.

File: src/popup/sections/ApiModes.jsx [hunk 20]
```diff
@@ -142,16 +143,18 @@ export function ApiModes({ config, updateConfig }) {
     }
   }, [])
 
-  const updateWhenApiModeDisabled = (apiMode) => {
-    if (isApiModeSelected(apiMode, config))
-      updateConfig({
+  const getSelectionPatchWhenApiModeDisabled = (apiMode) => {
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/sections/ApiModes.jsx]: Fixed in 7e97af5. The selection fallback check now reuses isApiModeSelected with apiMode cleared, so legacy ollamaModel/azureOpenAi values are compared through the existing canonical resolution while the raw modelName remains persisted. Tests cover Ollama, Azure, disabling the fallback itself, and duplicate custom fallbacks; the built Chromium extension also preserves ollamaModel/llama4 when disabling the temporary ChatGPT override.

File: src/popup/api-mode-config-utils.mjs [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+    knownApiModeDefaultIds,
+  }
+}
+
+export function getSelectionPatchWhenApiModeDisabled(apiMode, apiModes, config) {
+  if (!isApiModeSelected(apiMode, config)) return {}
+
+  const fallbackConfig = { ...config, apiMode: null }
+  const hasFallback = apiModes.some((candidate) => isApiModeSelected(candidate, fallbackConfig))
+  const disabledModeIsFallback = isApiModeSelected(apiMode, fallbackConfig)
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:30]: No change: disabledModeIsFallback is reachable. With apiMode cleared, isApiModeSelected intentionally falls back to config.modelName and canonicalizes legacy Ollama/Azure values. For example, modelName=ollamaModel plus ollamaModelName=llama4 matches the ollamaModel-llama4 row, so disabling that fallback must clear it to customModel. The unit test getSelectionPatchWhenApiModeDisabled clears a disabled legacy fallback exercises this exact branch and passes; removing the guard would retain a mode the user just disabled.

File: src/popup/api-mode-config-utils.mjs [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+  return {
+    modelName: hasFallback && !disabledModeIsFallback ? config.modelName : 'customModel',
+    apiMode: null,
+  }
+}
```
[the remark this answers is not in the document]
Comment (PeterDaveHello) [reply src/popup/api-mode-config-utils.mjs:36]: No change. This would not select the remaining provider: modelName stores only the canonical model name, while provider identity exists only in apiMode. The suggestion clears apiMode, so two same-name provider rows remain ambiguous and getUniquelySelectedApiModeIndex returns -1; preserving modelName cannot pin the selection to the sibling. The pre-PR helper also reset to customModel when disabling the row whose canonical name equaled the fallback, so this is not a regression from the fix. The new duplicate test covers the distinct legacy behavior that must remain: disabling an unrelated temporary override preserves the pre-existing name fallback even when that fallback is ambiguous.
